vllm/spec_decode/batch_expansion.py

```diff
@@ -196,21 +188,55 @@ def _contract_batch(
             all_hidden_states = None
 
         # Rule out prefills that produce no tokens.
-        non_spec_indices = [
+        regular_sampling_indices = [
             idx for idx in non_spec_indices
             if contracted_seq_group_metadata_list[idx].do_sample
         ]
+        has_prompt_log = any((sg.sampling_params.prompt_logprobs
+                              and sg.sampling_params.prompt_logprobs > 0)
+                             for sg in contracted_seq_group_metadata_list)
+        # When prompt logprobs is enabled, lens of returned tensors go from
+        # n_sampled (requests with do_sample=True) to n_prompt+n_prefills.
+        # We adjust stride accordingly to get the generated tokens and
+        # their probs, but pass on prompt_logprobs as is.
+        if not has_prompt_log:
+            # When prompt logprobs are not to be returned,
+            # we can ignore non-terminal chunks.
+            non_spec_indices = regular_sampling_indices
+
         if len(non_spec_indices):
+            if has_prompt_log:
+                # Add all terminal chunks sizes as well as decodes with no
+                # speculation to get out tokens and skip over prompt ones.
+                seq_meta = contracted_seq_group_metadata_list
+                nospec_sizes = torch.tensor([
```

Comment:
we don't ignore non-terminal chunks here, we actually have to pick their corresponding "output" token (which is always -1) so that it can be post-processed. These -1s are simply discarded later, but we're just complying with current state of post-processing. 
Basically it's only needed to have matching number of num_input_request and num_outputs (tokens/probs).